core/src/main/java/com/linecorp/armeria/common/stream/ThreeElementFixedStreamMessage.java

```diff
@@ -0,0 +1,178 @@
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.annotation.UnstableApi;
+
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * A {@link FixedStreamMessage} that publishes three objects.
+ */
+@UnstableApi
+public class ThreeElementFixedStreamMessage<T> extends FixedStreamMessage<T> {
+
+    @Nullable
+    private T obj1;
+    @Nullable
+    private T obj2;
+    @Nullable
+    private T obj3;
+
+    private boolean inOnNext;
+    private volatile int demand;
+
+    /**
+     * Constructs a new {@link ThreeElementFixedStreamMessage} for the given objects.
+     */
+    protected ThreeElementFixedStreamMessage(T obj1, T obj2, T obj3) {
+        this.obj1 = obj1;
+        this.obj2 = obj2;
+        this.obj3 = obj3;
+    }
+
+    @Override
+    public final boolean isEmpty() {
+        return false;
+    }
+
+    @Override
+    public long demand() {
+        return demand;
+    }
+
+    @Override
+    final void cleanupObjects(@Nullable Throwable cause) {
+        if (obj1 != null) {
+            StreamMessageUtil.closeOrAbort(obj1, cause);
+            obj1 = null;
+        }
+        if (obj2 != null) {
+            StreamMessageUtil.closeOrAbort(obj2, cause);
+            obj2 = null;
+        }
+        if (obj3 != null) {
+            StreamMessageUtil.closeOrAbort(obj3, cause);
+            obj3 = null;
+        }
+    }
+
+    @Override
+    public void request(long n) {
+        final EventExecutor executor = executor();
+        if (executor.inEventLoop()) {
+            request0(n);
+        } else {
+            executor.execute(() -> request0(n));
+        }
+    }
+
+    private void request0(long n) {
+        if (obj3 == null) {
+            return;
+        }
+
+        if (n <= 0) {
+            onError(new IllegalArgumentException(
+                    "n: " + n + " (expected: > 0, see Reactive Streams specification rule 3.9)"));
+            return;
+        }
+
+        if (inOnNext) {
+            // re-entrance
+            if (n == 1) {
+                demand++;
+            } else {
+                demand += 2;
+            }
+            return;
+        }
+
+        if (n >= 3) {
+            // All elements will be consumed. No need to restore inOnNext
+            inOnNext = true;
+            if (obj1 != null) {
+                final T item = obj1;
+                obj1 = null;
+                onNext(item);
+            }
+            if (obj2 != null) {
+                final T item = obj2;
+                obj2 = null;
+                onNext(item);
+            }
+            if (obj3 != null) {
+                final T item = obj3;
+                obj3 = null;
+                onNext(item);
+                onComplete();
+            }
+        } else {
+            demand += n;
+            if (obj1 != null) {
+                final T item = obj1;
+                obj1 = null;
+                inOnNext = true;
```

Comment:
Don't we need to use try and finally?

Reply:
All `Subscriber`'s methods in `FixedStreamMessage` are wrapped by try-catch statement. Only the fatal exceptions are rethrown. I thought there is nothing we can do if a fatal exception is raised.